### DISCUSSION TITLE
Words registers

### DISCUSSION THREAD

**Discussion by Ridente** [pymodbus-dev/pymodbus#1787] [General]
Hello
 I have managed to read and write within these devices (please see links). I am able to read bytes characters (2 address)
I would like to ask two questions:

1)
I need to write a float point value ( 2 byte register or word)

Function codes Tables : page 1-1
Data Format : page 2-2 and 2-4

2)
I am unable to read a word register; the system responds with 'illegal address' . I suspect the address is being sent incorrectly. 
For example, on the same page mentioned earlier, 6-4, how can I read register '30503'?"

Thank so much for any help

**Comment by janiversen** [comment]
You miss something fundamentally, you do not read bytes you read registers and a register is 16 bit which is the same as a word.

**Comment by janiversen** [comment]
In respect of reading the address, you do not show what you try to do, so it's impossible to tel you what's wrong!

Did you read our documentation, that tells clearly how to use the function calls, we also provide ready to go examples, like simple_sync_client.py.

**Comment by Ridente** [comment]
I partially solved the problem. I was making a mistake in the byte sequence (it should be Big Endian) and I wasn't sending them correctly (minimal code) :


## Write Float
  `
 def write_regs(self, address: int, value: list):

            return self.client.write_registers(address=address, values=value, slave=self.slave)


 def write_float(self, address: int, value: float) :

        #  BIG ENDIAN
        float_bigendian = struct.pack("!f", value)

        #  bytes...
        float_bigendian_bytes = struct.unpack("!HH", float_bigendian)

       self.write_regs(address=address, value=float_bigendian_bytes)
`

## Read Float

`
        def read_output_registers(self, address: int, count: int):
               return self.client.read_holding_registers(address=address, count=count, slave=self.slave)

        def read_float(self, address: int) -> float:
               data = self.read_output_registers(address=address, count=2)

        bytes = BinaryPayloadDecoder.fromRegisters(data.registers, byteorder=Endian.BIG, wordorder=Endian.BIG)

        return bytes.decode_32bit_float()
`

The code works, reads, and writes correctly. The only issue, but I'm not sure if it's related to the program, is that I can't read from the address in the second column. For example, on page 6-4:

I can read and write with register 0536, but with the same code, I can't read from the adjacent register 30537 (read-only).
It return [] . I 'll post here the log

**Reply by janiversen** [reply]
But since you can read some addresses this is not a problem with pymodbus !

**Comment by Ridente** [comment]
Sure , thank you for your help
